Funções de pegar / devolver / pesquisar livros

heber364 opened the pull request:
### O que há de novo:
- Agora é possível usar o _id_ do _user_ do _firebase_ através _useSession_;

- Um state de livros  pego emprestados pelo usuário, assim como sua interface;

- 4 novas funções principais: **_borrowBook_ (pegar livro emprestado)**, **_returnBookUser_ (devolver livro)**, 
**reBorrowBook (pegar novamente livro emprestado)** e **_requestBooksUserFirebase_** que retorna os livros que usuário está ou já pegou emprestado anteriormente;

- Página dinâmica de livros, usando SSR;

- Função de pesquisa na página Buscar Livro;

### Coisas a serem analisados em próximas branchs:
- As páginas com SSR não atualizam até darmos refresh, quando fazemos alguma ação como pegar livro emprestado, devolver ou re-alugar;

- Responsividade, design de componentes e páginas e estruturação dos componentes;

- Função de pesquisa vai ser revisada e expandida para outras páginas;

heber364 committed:
Adicionado UserID do usuário ao useSession.

heber364 committed:
Conponente BookCard

heber364 committed:
Criando algumas funções

heber364 committed:
Corrigindo sidebar e +

- Ao clicar no link do sidebar, o mesmo é fechado apenas na versão mobile.

- Verificando status do livro e renderizando tag de acordo tal

heber364 committed:
Criando página dinâmica de livros

heber364 committed:
Página dinâmicas de livros e função de pegar emprestado

- Cada livro tem sua página usando SSR
- A interface dos livros possui agora _borrowedTo_ onde possuem dados como _user_ _startDate_ e _endDate_.
- A interface dos usuários possui agora _borrowedBooks_ que é um array de BookProps, onde fica listado todos os livros que usuário pegou emprestado

Guui-t committed:
Inserida funcionalidade de pesquisa e ajuste das imagens dos cards

Agora a barra de pesquisa funciona na página de pesquisa.

Guui-t committed:
Merge branch 'feat/book-barrow-function' of https://github.com/CT-junior/Biblioteca-CT into feat/book-barrow-function

Guui-t committed:
adcionado objectFit = cover

heber364 committed:
Página principal

- Nova interface UserBooksProps assim como
- Criando um função para buscar livros do usuário no firebase
- Atualizando zIndex no sidebar
- Página principal agora  mostra livros do usuário

heber364 committed:
update

heber364 committed:
Fix pegar livro emprestado novamente

heber364 committed:
Delete images

heber364 committed:
Recovery vercel logo

faakit commented on src/components/BooksDisplay/index.tsx:
```diff
@@ -196,10 +190,30 @@ export function BooksDisplay({
-                                    <TagLeftIcon as={MdOutlinePending} />
-                                    <TagLabel>Aguardando devolução</TagLabel>
-                                </Tag>
+                                {book.status === "devolvido" ? (
```
Você poderia passar a lógica do label direto no UpdateStatePopover, economizando as linhas desse componente, não tendo de passar o status como children (o que não é o melhor dos padrões)

faakit commented on src/pages/library/[id].tsx:
```diff
@@ -0,0 +1,168 @@
+            image: String(session?.user?.image),
+            borrowedBooks: null,
+        };
+        await borrowBook(book, user);
```
por que não passar loco session.user?
se for porque user pode ser undefined basta checar previamente:
```
if (!!session?.user){
  await borrowBook(book, session.user);
}
```

faakit commented on src/pages/library/[id].tsx:
```diff
@@ -0,0 +1,168 @@
+    id: string;
+}
+
+export const getServerSideProps: GetServerSideProps = async (context) => {
```
é realmente necessário receber isso direto do servidor? como essas páginas deverão sempre ser dinâmicas e não disponíveis publicamente não há um grande ponto em utilizar do SSR, não precisamos "esconder" nada do usuário aqui..

heber364 replied:
André, penso em fazer isso depois em algum feature.

faakit commented on src/store/books/actions.ts:
```diff
@@ -226,3 +221,224 @@ export const requestBooksFirebase = async () => {
+        });
+    } catch (err) {
+        store.update((s) => {
+            s.isLoading = true;
```
se der erro o loading vira infinito?

faakit commented on src/store/books/actions.ts:
```diff
@@ -226,3 +221,224 @@ export const requestBooksFirebase = async () => {
+        });
+    } catch (err) {
+        store.update((s) => {
+            s.isLoading = true;
```
loading infinito novamente

faakit commented:
Uma forma interessante de lidar com o firebase quando vc precisa fazer várias requisições (por exemplo atualizar um livro + o usuário na hr de pedir um livro emprestado) é utilizando a [TRANSACTION ](https://firebase.google.com/docs/firestore/manage-data/transactions), assim você garante que 2 pessoas não acabem disputando um mesmo recurso ocasionando uma falha na disponibilidade...

heber364 committed:
Fix BookDisplay

Guui-t reviewed (commented):
Como o código tem muita coisa que repete, seria melhor criar um componente pra estilizar e só mudar o estado dele, de acordo com o necessário

heber364 committed:
Fix comments

- Componentizando alguns componentes
- Mudando o nome de um tipo da interface de BooksUserProps
- Adicionando lógica de toast nos actions

heber364 committed:
Ajustes no código e componentes

 - Alterando tipo de dados na interface do user
- Criando um componente BookRow
- Criando um componente TableBooksUser
- Apagando bookDisplay ( a lógica de _array.map_ é passado nas páginas
- Consertando a "gambiarra" usada para criar a tabela do BooksDisplay

heber364 committed:
Fix comments

- Atualizando actions,: agora ao devolver um livro, ele é deletado da sua estante virtual pessoal;
- Adicionando title nas páginas;
- Toda lógica de verificação se o usuário está autenticado ou não, é passado no ProtectedRoutes.ts

-

heber364 committed:
Pesquisas

- Funções de pesquisar das páginas principal, library e library-manager finalizadas;

-

faakit commented on src/common/ProtectedRoutes.ts:
```diff
@@ -0,0 +1,48 @@
+            if (books.length === 0) requestBooksFirebase();
+            if (booksUser.length === 0)
+                requestBooksUserFirebase(session?.user?.id);
+            if (registries.length === 0) requestRegistriesFirebase();
```
não é melhor deixar q só as páginas que usam esses recursos chamem essas requests?

heber364 replied:
Eu pensei nisso para depois as páginas não demorarem ser renderizadas (primeira vez no caso), pq os dados já estarão carregados.

faakit replied:
porém a confiabilidade do dado vai diminuir consideravelmente (vai ficar desatualizado)

faakit commented on src/components/BookRow/Popover/index.tsx:
```diff
@@ -0,0 +1,103 @@
+export function Popover({ book }: UpdateStatePopoverProps) {
+    const { data: session } = useSession();
+    const [loading, setLoading] = useState(false);
+    const [loading2, setLoading2] = useState(false);
```
creio que seja melhor usar apenas um loading compartilhado entre eles, afinal se você clica em um botão, não deveria clicar no outro.

heber364 replied:
Boa, faz sentido, vou alterar.

faakit commented on src/pages/library-manager.tsx:
```diff
@@ -150,7 +138,7 @@ const LibraryManager: NextPage = () => {
                 })}
             </TableLibraryManager>
-            <Pagination />
+            {/* <Pagination /> */}
```
remover código comentado

faakit commented on src/pages/library/[id].tsx:
```diff
@@ -0,0 +1,151 @@
+    const { isLoading } = useBooks();
+    const handleBorrowBook = async () => {
+        await borrowBook(book, session.user);
+        router.reload();
```
o motivo do reload é pra carregar o quê? talvez seja interessante dar os refresh's necessários sem recarregar a página (é esquisito ver a página recarregar literalmente)

heber364 replied:
Quando pego emprestado o livro, as informações sobre o livro não alteram. Ai eu forço renderização. Obviamente não é a melhor prática, mas enquanto não deixei como estado, não é possível.

heber364 replied:
Mas vou fazer isso numa feature futura.

heber364 committed:
Fix comments
